polynote-server/src/main/scala/polynote/server/SharedNotebook.scala

```diff
@@ -0,0 +1,285 @@
+  * version X + 3 and send that transformed edit to client A (which will also apply any necessary transformations locally).
+  *
+  * Thus, the SharedNotebook must track its global version, the current notebook, and low watermark of the highest global version
+  * that all clients have acknowledged. It must also keep a history buffer of all edits between the low watermark and
+  * the current global version. Clients can periodically send a status message to the server indicating the current-known
+  * global version and the current local version, in order to allow the server to discard edit history.
+  *
+  * Clients must track the currently known global version, which updates whenever the client receives a foreign edit,
+  * and the highest local version which has been acknowledged by the server (either by sending an edit based on it,
+  * or through an acknowledgement message). It has to keep its local edit history between that known-acknowledged
+  * version and its current version, to allow rebasing new foreign edits.
+  *
+  * It sounds a lot more complicated than it really is.
+  */
+trait SharedNotebook[F[_]] {
+
+  def path: String
+
+  /**
+    * Open a reference to this shared notebook.
+    *
+    * @param name A string identifying who is opening the notebook (i.e. their username or email)
+    * @param oq   A queue to which messages to the subscriber (such as notifications of foreign updates) will be submitted
+    * @return A [[NotebookRef]] which the caller can use to send updates to the shared notebook
+    */
+  def open(name: String, oq: Enqueue[F, Message]): F[NotebookRef[F]]
+
+  def versions: Stream[F, (Int, Notebook)]
+}
+
+
+class IOSharedNotebook(
+  val path: String,
+  ref: SignallingRef[IO, (GlobalVersion, Notebook)],            // the Int is the global version, which can wrap around back to zero if necessary
+  kernelRef: Ref[IO, Option[Kernel[IO]]],
+  updates: Queue[IO, Option[(SubscriberId, NotebookUpdate, Deferred[IO, GlobalVersion])]],   // the canonical series of edits
+  updatesTopic: Topic[IO, Option[(GlobalVersion, SubscriberId, NotebookUpdate)]],  // a subscribe-able channel for watching updates,
+  kernelFactory: KernelFactory[IO],
+  outputMessages: Topic[IO, Message],
+  kernelLock: Semaphore[IO]
+)(implicit contextShift: ContextShift[IO]) extends SharedNotebook[IO] {
+
+  private val updateBuffer = new VersionBuffer[NotebookUpdate]
+
+  private val statusUpdates = Publish.PublishTopic(outputMessages).contramap[KernelStatusUpdate](KernelStatus(ShortString(path), _))
+
+  // listen to the stream of updates and apply them in order, each one incrementing the global version
+  updates.dequeue.unNoneTerminate.zipWithIndex.evalMap {
+    case ((subscriberId, update, versionPromise), version) =>
+      val newGlobalVersion = (version % Int.MaxValue).toInt
+      updateBuffer.add(newGlobalVersion, update)
```

Comment:
Hmm, should we consider adding the update to the `updateBuffer` inside `applyUpdate`? That way if it failed to apply for some reason we wouldn't add it (or handle that somehow)?

Reply:
Sure, makes sense.